### PR TITLE
Rewrite ForceUnwrappingRule using SwiftSyntax

### LINKED ISSUE #2915: force_unwrapping is not fired for trailing closures syntax
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

`force_unwrapping` is not fired for trailing closures syntax
```swift
let value = [1, 3, 5, 6].first { $0.isMultiple(of: 2) }!
```

Same case but with parentheses around works fine
```swift
let value = [1, 3, 5, 6].first({ $0.isMultiple(of: 2) })!
```

##### Complete output when running SwiftLint, including the stack trace and command used

E.g. for this case `superfluous_disable_command` rule is fired
```swift
let value = [1, 3, 5, 6].first { $0.isMultiple(of: 2) }! // swiftlint:disable:this force_unwrapping
```

```bash
$ swiftlint lint

Superfluous Disable Command Violation: SwiftLint rule 'force_unwrapping' did not trigger a violation in the disabled region. Please remove the disable command. (superfluous_disable_command)
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?
0.35.0
* Installation method used (Homebrew, CocoaPods, building from source, etc)?
CocoaPods
* Paste your configuration file:
```yml
opt_in_rules:
  - force_unwrapping
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
no
* Which Xcode version are you using (check `xcode-select -p`)?
Version 11.1 (11A1027)
* Do you have a sample that shows the issue? Run `echo "let value = [1, 3, 5, 6].first { $0.isMultiple(of: 2) }!" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
// This triggers a violation:
let value = [1, 3, 5, 6].first { $0.isMultiple(of: 2) }!
```

### PR DESCRIPTION
**Pull request by marcelofabri**
Fixes #2915

### CODE REVIEW

File: Source/SwiftLintFramework/Models/Location.swift [hunk 1]
```diff
@@ -60,9 +60,9 @@ public struct Location: CustomStringConvertible, Comparable, Codable {
     /// - parameter position: The absolute position returned from SwiftSyntax.
     public init(file: SwiftLintFile, position: AbsolutePosition) {
         self.file = file.path
-        if let sourceLocation = file.locationConverter?.location(for: position) {
-            line = sourceLocation.line
-            character = sourceLocation.column
+        if let lineAndCharacter = file.stringView.lineAndCharacter(forByteOffset: ByteCount(position.utf8Offset)) {
```
Comment (jpsim) [inline Source/SwiftLintFramework/Models/Location.swift:63]: @marcelofabri why this change? Did you benchmark it as being faster?

Comment (jpsim) [reply Source/SwiftLintFramework/Models/Location.swift:63]: And why not delegate to this init? https://github.com/realm/SwiftLint/blob/79347a17286476509ff79cf92f959e42597d4b56/Source/SwiftLintFramework/Models/Location.swift#L45

Comment (marcelofabri) [reply Source/SwiftLintFramework/Models/Location.swift:63]: I didn’t benchmark it, but the existing code was giving wrong column numbers for lines with UTF-8 characters. The added example was one case.

Comment (marcelofabri) [reply Source/SwiftLintFramework/Models/Location.swift:63]: (I should’ve delegated it to the other init though)

File: Source/SwiftLintFramework/Rules/Idiomatic/ForceUnwrappingRule.swift [hunk 2]
```diff
@@ -39,151 +38,44 @@ public struct ForceUnwrappingRule: OptInRule, ConfigurationProviderRule {
...
-            return false
-        }
+private final class ForceUnwrappingVisitor: SyntaxVisitor, ViolationsSyntaxVisitor {
+    private(set) var violationPositions: [AbsolutePosition] = []
 
-        // range is type annotation of declaration
-        return true
+    override func visitPost(_ node: ForcedValueExprSyntax) {
+        violationPositions.append(node.exclamationMark.positionAfterSkippingLeadingTrivia)
     }
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/Idiomatic/ForceUnwrappingRule.swift:80]: chefs kiss gif

File: Source/SwiftLintFramework/Models/Location.swift (hunk 1 above)
Comment (jpsim) [reply Source/SwiftLintFramework/Models/Location.swift:63]: Giving our caching of `StringView`s, it may actually be faster and more accurate than SwiftSyntax.

I'd encourage you to benchmark it in SwiftLint and to file a bug report towards SwiftSyntax if there's a correctness bug.

Comment (jpsim) [reply Source/SwiftLintFramework/Models/Location.swift:63]: > (I should’ve delegated it to the other init though)

Good opportunity for a follow-up PR 😄

Comment (marcelofabri) [reply Source/SwiftLintFramework/Models/Location.swift:63]: I’ll make that change and open an issue with SwiftSyntax later today 🙏

File: Source/SwiftLintFramework/Rules/Idiomatic/ForceUnwrappingRule.swift (hunk 2 above)
Comment (SimplyDanny) [reply Source/SwiftLintFramework/Rules/Idiomatic/ForceUnwrappingRule.swift:80]: Same reaction ... 😂

File: Source/SwiftLintFramework/Models/Location.swift (hunk 1 above)
Comment (marcelofabri) [reply Source/SwiftLintFramework/Models/Location.swift:63]: I think we were using this wrong. From `SourceLocation`:

```
/// The UTF-8 byte offset from the beginning of the line where this location
  /// resides. 1-based.
  public var column: Int? {
    return compLoc?.column
  }
```

As far as I can tell, we expect `Location`'s `character` to be a UTF-16 offset, so this would explain the behavior I was seeing. Does this make sense?

Comment (jpsim) [reply Source/SwiftLintFramework/Models/Location.swift:63]: Hmmm, I expect it to be the logical character offset on the line so that Xcode renders it at the correct location.
